frame/examples/tasks-example/src/lib.rs

```diff
@@ -0,0 +1,135 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use frame_support::{dispatch::DispatchResult, traits::Task};
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+use sp_runtime::DispatchError;
+
+#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+pub enum ExampleTask {
+	Increment,
+	Decrement,
+}
+
+pub trait PalletTask<T: Config>: Task {
+	type Config: frame_system::Config;
+	fn is_valid(&self, config: &Self::Config) -> bool;
+	fn run(&self, config: &Self::Config) -> Result<(), DispatchError>;
+}
+
+impl Task for ExampleTask {
+	type Enumeration = std::vec::IntoIter<ExampleTask>;
+
+	const TASK_INDEX: usize = 0;
+
+	fn enumerate() -> Self::Enumeration {
+		vec![ExampleTask::Increment, ExampleTask::Decrement].into_iter()
+	}
+
+	fn is_valid(&self) -> bool {
+		unimplemented!()
+	}
+
+	fn run(&self) -> Result<(), DispatchError> {
+		unimplemented!()
+	}
+}
+
+impl<T: Config> PalletTask<T> for ExampleTask {
+	type Config = T;
+
+	fn is_valid(&self, _config: &Self::Config) -> bool {
+		let value = Value::<T>::get().unwrap();
+		match self {
+			ExampleTask::Increment => value < 255,
+			ExampleTask::Decrement => value > 0,
+		}
+	}
+
+	fn run(&self, _config: &Self::Config) -> Result<(), DispatchError> {
+		match self {
+			ExampleTask::Increment => {
+				// Increment the value and emit an event
+				let new_val = Value::<T>::get().unwrap().checked_add(1).ok_or("Value overflow")?;
+				Value::<T>::put(new_val);
+				Pallet::<T>::deposit_event(Event::Incremented { new_val });
+			},
+			ExampleTask::Decrement => {
+				// Decrement the value and emit an event
+				let new_val = Value::<T>::get().unwrap().checked_sub(1).ok_or("Value underflow")?;
+				Value::<T>::put(new_val);
+				Pallet::<T>::deposit_event(Event::Decremented { new_val });
+			},
+		}
+		Ok(())
+	}
+}
```

Comment:
So I know that to write `is_valid` and `run` we typically will need access to storage (please correct me if I'm wrong about this!). With our currently proposed version of the `Task` trait, there is no generic `T: Config`, and my understanding is we couldn't really add this directly as it would make `Task` no longer general.

So my thought was in the expansion of the macro we could do something like the above where we create some wrapper trait like `PalletTask<T: Config>: Task` and then in it's impl, shadow any methods that actually require use of `T: Config` or other bounds we may need. I _think_ the above is correct, though I have not tested it directly

Another approach might be to have an associated type called `Config` on `Task` but I don't think we'd be able to bind it by anything so that probably wouldn't work.

I've never really done this shadowing technique before so I'm also not sure about the soundness of simply marking `is_value` and `run` as unreachable and then shadowing them.

As a test of this, I chose having a simple `u8` value in storage, with the ability to `increment` or `decrement` it, where overflowing or underflowing the `u8` from either of these will result in a `DispatchError`. This allows me to have something non-trivial to test in my `is_valid` method. I then implemented `increment` and `decrement` both as task variants _and_ as simple calls. So in this case the task is a bit redundant but I figured this would be useful for testing.

Also although I did this in `frame/examples`, this is just so I can test something locally for a while.

Would love some feedback @gupnik @kianenigma @bkchr

Reply:
Actually, I would think that each pallet would optionally expose a Task<T> that would provide the details of tasks managed by that pallet. Eventually, `construct_runtime` would aggregate all of these into something like a `RuntimeTask`.

Reply:
Indeed. Similar to `Call`.